File: evo_motion/model/environment.py
from typing import Callable, List

from pydantic import BaseModel
from torch import Tensor

from evo_motion.controller import Controller

from .engine import Engine
from .item import Item


class Step(BaseModel):
    state: Tensor
    reward: float
    done: bool

    class Config:
        arbitrary_types_allowed = True


class EnvironmentConfig(BaseModel):
    items: List[Item]
    controllers: List[Controller]
    engine: Engine
    compute_step: Callable[[List[Item]], Step]
    state_space: List[int]
    action_space: List[int]
    is_continuous: bool

    class Config:
        arbitrary_types_allowed = True


class Environment:
    def __init__(self, config: EnvironmentConfig):
        self.__engine = config.engine

        self.__items = config.items

        for item in self.__items:
            self.__engine.add_item(item)

        self.__controllers = config.controllers

        self.__state_space = config.state_space
        self.__action_space = config.action_space
        self.__is_continuous = config.is_continuous

        self.__compute_step = config.compute_step

    def step(self, action: Tensor) -> Step:
        for c in self.__controllers:
            c.on_input(action)

        self.__engine.step()

        return self.__compute_step(self.__items)

    def reset(self) -> Step:
        return self.__compute_step(self.__items)

File: src/model/environment.cpp
//
// Created by samuel on 18/12/22.
//

#include "environment.h"

Environment::Environment(const std::vector<int64_t> &state_space,
                         const std::vector<int64_t> &action_space,
                         bool continuous) :
    curr_device(torch::kCPU),
    state_space(state_space), action_space(action_space),
    continuous(continuous),
    m_collision_configuration(new btDefaultCollisionConfiguration()),
    m_dispatcher(new btCollisionDispatcher(m_collision_configuration)),
    m_broad_phase(new btDbvtBroadphase()),
    m_constraint_solver(new btSequentialImpulseConstraintSolver()),
    m_world(new btDiscreteDynamicsWorld(
        m_dispatcher,
        m_broad_phase,
        m_constraint_solver,
        m_collision_configuration)) {

    m_world->setGravity(btVector3(0, -9.8f, 0));

}

void Environment::add_item(Item item) {
    m_world->addRigidBody(item.get_body());
}

step Environment::do_step(const torch::Tensor &action, float delta) {
    for (const auto &c: get_controllers())
        c->on_input(action);

    m_world->stepSimulation(delta);

    return compute_step();
}

step Environment::reset() {
    reset_engine();
    return compute_step();
}

std::vector<int64_t> Environment::get_state_space() {
    return state_space;
}

std::vector<int64_t> Environment::get_action_space() {
    return action_space;
}

bool Environment::is_continuous() const {
    return continuous;
}

void Environment::to(torch::DeviceType device) {
    curr_device = device;
}
